File: src/main/java/com/learnspark/notification/entity/ReminderSetting.java
package com.learnspark.notification.entity;

import com.learnspark.common.entity.BaseEntity;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.Id;
import jakarta.persistence.PrePersist;
import jakarta.persistence.Table;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.time.LocalDateTime;
import java.time.LocalTime;

/**
 * 提醒设置实体，对应 reminder_settings 表。
 *
 * <p>1:1 关联 users 表，主键即 user_id。
 * 每用户一条记录，记录提醒时间、时区、开关与上次发送时间。
 */
@Entity
@Table(name = "reminder_settings")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class ReminderSetting extends BaseEntity {

    /** 用户 ID（主键 + 外键 → users.id） */
    @Id
    @Column(name = "user_id", columnDefinition = "VARCHAR(36)")
    private String userId;

    /** 接收提醒的邮箱 */
    @Column(name = "email", nullable = false, length = 255)
    private String email;

    /** 每日提醒时间（如 08:00:00） */
    @Column(name = "reminder_time", nullable = false)
    private LocalTime reminderTime;

    /** 时区（如 Asia/Shanghai） */
    @Column(name = "timezone", length = 50)
    private String timezone;

    /** 是否启用提醒 */
    @Column(name = "enabled")
    private Boolean enabled;

    /** 通知渠道（JSON，预留扩展，如 ["email","browser"]） */
    @Column(name = "channels", columnDefinition = "JSON")
    private String channels;

    /** 上次发送时间（防重复发送） */
    @Column(name = "last_sent_at")
    private LocalDateTime lastSentAt;

    /** 持久化前填充默认值 */
    @PrePersist
    void fillDefaults() {
        if (this.timezone == null) {
            this.timezone = "Asia/Shanghai";
        }
        if (this.enabled == null) {
            this.enabled = true;
        }
    }
}

File: learnspark-kmp/server/src/main/kotlin/com/learnspark/server/domain/entity/ReminderSetting.kt
package com.learnspark.server.domain.entity

import jakarta.persistence.Column
import jakarta.persistence.Entity
import jakarta.persistence.EnumType
import jakarta.persistence.Enumerated
import jakarta.persistence.Id
import jakarta.persistence.Table
import java.time.Instant
import java.time.LocalTime

/**
 * R3b：用户自定义提醒规则。
 *
 * triggerTime 是 LocalTime（HH:mm:ss），每自然日同一时刻触发。
 * repeatPattern 控制触发频率：
 *  - once: 仅触发一次（next_fire_at 触发后置 null）
 *  - daily: 每日
 *  - weekdays: 周一到周五（忽略 weekdayMask）
 *  - weekly:  按 weekdayMask 选定的星期几触发
 *
 * weekdayMask 是位掩码：bit0=Mon, bit1=Tue, ..., bit6=Sun
 * 默认 127 = 全选
 */
@Entity
@Table(name = "reminder_settings")
class ReminderSetting(
    @Id
    var id: String = "",

    @Column(name = "user_id", nullable = false)
    var userId: String = "",

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    var type: ReminderType = ReminderType.CUSTOM,

    @Column(nullable = false)
    var title: String = "",

    @Column(columnDefinition = "TEXT")
    var message: String? = null,

    @Column(name = "target_id")
    var targetId: String? = null,

    @Column(name = "trigger_time", nullable = false)
    var triggerTime: LocalTime = LocalTime.of(9, 0),

    @Enumerated(EnumType.STRING)
    @Column(name = "repeat_pattern", nullable = false)
    var repeatPattern: RepeatPattern = RepeatPattern.DAILY,

    @Column(name = "weekday_mask", nullable = false)
    var weekdayMask: Int = 127,

    @Column(name = "next_fire_at")
    var nextFireAt: Instant? = null,

    @Column(nullable = false)
    var enabled: Boolean = true,

    @Column(nullable = false)
    var version: Long = 0,

    @Column(name = "created_at", insertable = false, updatable = false)
    var createdAt: Instant? = null,

    @Column(name = "updated_at", insertable = false, updatable = false)
    var updatedAt: Instant? = null,
)

enum class ReminderType { TASK_DUE, CHECK_IN, CUSTOM }
enum class RepeatPattern { ONCE, DAILY, WEEKDAYS, WEEKLY }
